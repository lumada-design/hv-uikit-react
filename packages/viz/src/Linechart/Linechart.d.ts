import { HvChartProps } from "../Chart";

export interface HvLinechartProps extends HvChartProps {
  /**
   * Sets the type of graph.
   */
  type?: "line" | "area" | "stack";
  /**
   * Defines it should present a range slider.
   */
  rangeSlider?: boolean;
}

export default function HvLinechart(props: HvLinechartProps): JSX.Element | null;
