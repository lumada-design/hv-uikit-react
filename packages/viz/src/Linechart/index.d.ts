export { default } from "./Linechart";
export * from "./Linechart";
