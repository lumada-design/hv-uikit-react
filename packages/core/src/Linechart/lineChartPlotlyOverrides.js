/*
 * Copyright 2019 Hitachi Vantara Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import clone from "lodash/cloneDeep";

/**
 * Set the property if it is null.
 * @param object
 * @param property
 * @param value
 */
const setterIfNil = (object, property, value) => {
  // eslint-disable-next-line no-param-reassign
  object[property] = object[property] || value;
};

/**
 * Auxiliary functions to set the data in order to follow the DS guideline.
 *
 */
export const setData = (inputData, type) => {
  const data = inputData;

  data.forEach(trace => {
    setterIfNil(trace, "hoverinfo", "none");
    setterIfNil(trace, "mode", "lines");

    switch (type) {
      case "area":
        setterIfNil(trace, "fill", "tonexty");
        setterIfNil(trace, "type", "scatter");
        break;
      case "line":
        setterIfNil(trace, "type", "line");
        break;
      case "stack":
        setterIfNil(trace, "fill", "tonexty");
        setterIfNil(trace, "type", "scatter");
        setterIfNil(trace, "stackgroup", "one");
        break;
      default:
        setterIfNil(trace, "type", "line");
    }
  });

  return data;
};

export const setLayout = inputLayout => {
  const layout = inputLayout === undefined ? {} : clone(inputLayout);

  return layout;
};
