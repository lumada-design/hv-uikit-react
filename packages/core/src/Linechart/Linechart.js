import React, { useMemo } from "react";
import PropTypes from "prop-types";
import { withStyles } from "@material-ui/core";
import Chart from "../Chart";
import { applyDataDefaults, applyLayoutDefaults } from "./lineChartPlotlyOverrides";
import styles from "./styles";

/**
 * A line chart or line plot or line graph is a type of chart which displays information as a series of data points
 * called 'markers' connected by straight line segments. It is a basic type of chart common in many fields.
 *
 * Our implementation uses as base Plotly. If you have a specific case
 * that we don't cover directly, the Plotly [documentation](https://plotly.com/javascript/) is a good starting point.
 */
const Linechart = ({
  id,
  classes,
  title,
  subtitle,
  data,
  layout,
  config,
  tooltipType = "multiple",
  type = "line",
  xAxisTitle,
  yAxisTitle,
  rangeSlider = false,
  ...others
}) => {
  /* Values derived from props */

  const chartData = useMemo(() => applyDataDefaults(data, type), [data, type]);
  const chartLayout = useMemo(
    () => applyLayoutDefaults(layout, rangeSlider),
    [layout, rangeSlider]
  );

  return (
    <Chart
      id={id}
      classes={classes}
      title={title}
      subtitle={subtitle}
      xAxisTitle={xAxisTitle}
      yAxisTitle={yAxisTitle}
      data={chartData}
      layout={chartLayout}
      config={config}
      tooltipType={tooltipType}
      {...others}
    />
  );
};

Linechart.propTypes = {
  /**
   * An Id passed on to the component
   */
  id: PropTypes.string,
  /**
   * A Jss Object used to override or extend the styles applied.
   */
  classes: PropTypes.instanceOf(Object),
  /**
   * Title of the chart.
   */
  title: PropTypes.string,
  /**
   * Subtitle of the chart.
   */
  subtitle: PropTypes.string,
  /**
   * Plotly data object (see https://plot.ly/javascript/reference/).
   */
  data: PropTypes.arrayOf(PropTypes.instanceOf(Object)).isRequired,
  /**
   * Plotly layout object (see https://plot.ly/javascript/reference/#layout).
   */
  layout: PropTypes.instanceOf(Object),
  /**
   * Plotly config object (see https://plot.ly/javascript/configuration-options/).
   */
  config: PropTypes.instanceOf(Object),
  /**
   * Defines if should use a single or multiline tooltip.
   */
  tooltipType: PropTypes.oneOf(["single", "multiple"]),
  /**
   * Sets the type of graph.
   */
  type: PropTypes.oneOf(["line", "area", "stack"]),
  /**
   * Defines the title of the X axis.
   */
  xAxisTitle: PropTypes.string,
  /**
   * Defines the title of the Y axis.
   */
  yAxisTitle: PropTypes.string,
  /**
   * Defines it should present a range slider.
   */
  rangeSlider: PropTypes.bool,
};

export default withStyles(styles, { name: "HvLinechart" })(Linechart);
