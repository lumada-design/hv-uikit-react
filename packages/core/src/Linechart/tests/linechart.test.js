/*
 * Copyright 2019 Hitachi Vantara Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* eslint-env jest */

import React from "react";
import { mount, shallow } from "enzyme";

import HvProvider from "../../Provider";
import LinechartWithStyles from "../index";
import Linechart from "../Linechart";
import {setData} from "../lineChartPlotlyOverrides";

describe("Linechart withStyles", () => {
  let wrapper;

  const trace1 = {
    x: [2300, 1000, 8500],
    y: ["January", "February", "March"],
    name: "Sales Target"
  };

  const data = [trace1];
  const layout = {};

  beforeEach(() => {
    wrapper = shallow(
      <HvProvider>
        <LinechartWithStyles data={data} layout={layout} />
      </HvProvider>
    );
  });

  it("should be defined", () => {
    expect(wrapper).toBeDefined();
  });

  it("should render correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render the Linechart", () => {
    wrapper = mount(
      <HvProvider>
        <LinechartWithStyles data={data} layout={layout} />
      </HvProvider>
    );
    const linechart = wrapper.find(Linechart);
    expect(linechart.length).toBe(1);
  });

  it("should set default data for area", () => {
    const traceTest = {
      x: [2300, 1000, 8500],
      y: ["January", "February", "March"],
      name: "Sales Target"
    };

    const overwrittenData = setData([traceTest], "area");
    expect(overwrittenData[0].mode).toBe("lines");
    expect(overwrittenData[0].hoverinfo).toBe("none");
    expect(overwrittenData[0].fill).toBe("tonexty");
    expect(overwrittenData[0].type).toBe("scatter");
  });

  it("should set default data for line", () => {
    const traceTest = {
      x: [2300, 1000, 8500],
      y: ["January", "February", "March"],
      name: "Sales Target"
    };
    const overwrittenData = setData([traceTest], "line");
    expect(overwrittenData[0].mode).toBe("lines");
    expect(overwrittenData[0].hoverinfo).toBe("none");
    expect(overwrittenData[0].type).toBe("line");
  });

  it("should set default data for stack", () => {
    const traceTest = {
      x: [2300, 1000, 8500],
      y: ["January", "February", "March"],
      name: "Sales Target"
    };
    const overwrittenData = setData([traceTest], "stack");
    expect(overwrittenData[0].mode).toBe("lines");
    expect(overwrittenData[0].hoverinfo).toBe("none");
    expect(overwrittenData[0].fill).toBe("tonexty");
    expect(overwrittenData[0].type).toBe("scatter");
    expect(overwrittenData[0].stackgroup).toBe("one");
  });

  it("shouldn't override existing data", () => {
    const traceTest = { hoverinfo: "xx" };
    const overwrittenData = setData([traceTest]);
    expect(overwrittenData[0].hoverinfo).toBe("xx");
  });
});
