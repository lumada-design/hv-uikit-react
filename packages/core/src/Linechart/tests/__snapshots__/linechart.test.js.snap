// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Linechart withStyles should render correctly 1`] = `
<JssProvider
  generateClassName={[Function]}
>
  <HvProvider
    changeTheme={[Function]}
    router={null}
    theme={null}
    uiKitTheme="dawn"
  >
    <WithStyles(Linechart)
      data={
        Array [
          Object {
            "name": "Sales Target",
            "x": Array [
              2300,
              1000,
              8500,
            ],
            "y": Array [
              "January",
              "February",
              "March",
            ],
          },
        ]
      }
      layout={Object {}}
    />
  </HvProvider>
</JssProvider>
`;
